### ISSUE TITLE
Web UI looks broken during rolling updates

### ISSUE THREAD

**Issue by roobre** [mastodon/mastodon#33188]
### Steps to reproduce the problem

1. Run mastodon
2. Start rolling an update, e.g. in docker-compose or Kubernetes environments
3. While the update is being deployed, styles will not be applied to the website due to a nonce mismatch (unless you're really lucky)

### Expected behaviour

The website should still look good during rolling updates

### Actual behaviour

The website is broken and unusable as styles are not loaded due to a nonce mismatch

### Detailed description

In a rolling update situation, for a non-zero period of time (typically between 10 seconds and 1 minute), both old and new version of mastodon are serving HTTP requests. It is possible, and even likely, that a request to the UI is handled by version A, and a request to `/packs/css/...` is handled by version B.

If this happens, the styles for the new version served by B won't match the nonce set in the `content-security-policy` header for the response served by A.

### Mastodon instance

owo.cafe

### Mastodon version

v4.3.2

### Browser name and version

Firefox 134.0b5 (64-bit)

### Operating system

Arch Linux

### Technical details

During the upgrade, I was able to see logs in the browser console complaining about the nonce mismatch. Unfortunately I did not take a screenshot :(

**Comment by renchap** [comment]
This is weird, the nonces should not be used for remote resources (like `/packs/css`), but only for inline CSS/JS blocks. 

There can be 404s during an upgrade if you are using a Docker-based setup where `public/packs` only contain the current version, this is why you might want to either have some caching for those files in your web server, or copy the assets to another location that is used as the origin for `/packs`, similar to what we do here for Kubernetes setups: https://github.com/mastodon/chart/blob/main/templates/job-assets-copy.yaml

**Comment by roobre** [comment]
@renchap In all fariness, the `/packs/css` thing is something I guessed. I'm positive of seeing the nonce mismatch errors but I do not remember the URLs, so what you say is definitely possible.

I think caching is enough of a fix for the assets that go away during updating, or at least I'm not very concerned about those. 

I can try and reproduce the update process on my staging environment and try to grab the contents of the dev console. Would that help figuring out what is failing the nonce checks exactly?

**Comment by renchap** [comment]
Yes it would. But I do not think that the `nonce` message in the console will point to URLs, rather to specific blocks (`<style>` or `<script>`) in the page. In this case, please try to also report which blocks they point to.

**Comment by ClearlyClaire** [comment]
I do not think that this is an issue with nonces, but IIRC our use of Sub-Resource Integrity may lead to error messages that may obscure the following scenario:
- new web server mentions `/packs/css/foo-2`
- request hits old web server which only knows `/pack/css/foo-1`, raises a 404, the 404 response triggers SRI mismatch

**Comment by roobre** [comment]
Ran again through this to try and clarify a bit what's going on.

There are the messages I originally spotted when the site looked broken (without styles):

![Console log view showing content-security-policy violations](https://github.com/user-attachments/assets/75bcbf44-d509-42e6-adab-0e8e8bdca5e5)

Turns out these appear every time I reload the page. I'm not sure about why, but they seem to be unrelated to the issue. Seeing them while the page was broken made me jump to the conclusion of this being the issue.

I re-run the rolling update procedure to pay more attention this time and see what is actually broken:

![Image](https://github.com/user-attachments/assets/3658f982-9763-424a-ad0f-77768734e1f5)

And indeed as you mentioned, this seems to be way simpler: CSS packs change and old/new URLs return 404:

![Image](https://github.com/user-attachments/assets/8b826699-e2fd-4a6e-a09b-1a93bcaa3e15)

So yeah, nothing to do with CSPs as both of you said, sorry for the misleading report 😓 

Taking a look at the `cache-control` header, it seems like CSS styles have proper caching instructions:

![Image](https://github.com/user-attachments/assets/f4579c62-c96f-4825-984e-490a2b017683)

So I think I saw the web UI broken because of my tendency of shift-reload (without cache) when I'm doing changes. I don't think a normal user would ever see this, which is good!

I'll close this as there doesn't seem to be any bug, but rather an artifact of my cache-busting habits :P

Thank you both for your input!
